src/var/mutation_counting.jl

```diff
@@ -0,0 +1,152 @@
+# phylo/mutation_counting.jl
+# ==================
+#
+# Types and methods for counting different kinds of mutations between two
+# DNA or RNA sequences.
+#
+# Part of the Bio.Phylo module.
+#
+# This file is a part of BioJulia. License is MIT: https://github.com/BioJulia/Bio.jl/blob/master/LICENSE.md
+
+
+# Counting mutations
+# ------------------
+
+# Mutation types
+
+immutable Different end
+immutable Transition end
+immutable Transversion end
+
+
+"""
+    is_ambiguous_strict{T<:Nucleotide}(a::T, b::T)
+
+The strictest test of ambiguity, if either nucleotide is not
+A, T, G, or C, then this function returns true.
+"""
+@inline function is_ambiguous_strict{T<:Nucleotide}(a::T, b::T)
+    return isambiguous(a) || isambiguous(b)
+end
+
+typealias alphabets Union{DNAAlphabet{2}, DNAAlphabet{4}, RNAAlphabet{2}, RNAAlphabet{4}}
+
+"""
+    count_mutations(seq1::BioSequence{A}, seq2::BioSequence{A}, kind::Type{Different})
+
+Count the number of differences between two DNA sequences.
+Returns a tuple of the number of sites that are different, and the number of sites
+considered (sites with any ambiguity characters are not considered).
+"""
+function count_mutations{A<:alphabets}(seq1::BioSequence{A},
+    seq2::BioSequence{A},
+    t::Type{Different})
+
+    @assert length(seq1) == length(seq2)
+    nDifferences = 0
+    nSites = length(seq1)
+
+    for (a, b) in zip(seq1, seq2)
+        if is_ambiguous_strict(a, b)
+            nSites -= 1
+            continue
+        end
+        if a != b
```

Comment:
I see what you mean. I hesitate to add a run-time decision on the mutation type though. Would a generated function work for this purpose?